rcshist.1
.\"
.\" Copyright (c) 2001 Ian Dowse <iedowse@maths.tcd.ie>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer
.\"    in this position and unchanged.
.\" 2. The name of the author may not be used to endorse or promote products
.\"    derived from this software without specific prior written permission
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\" $Id: rcshist.1,v 1.4 2025/01/01 23:03:27 tom Exp $
.\"
.TH RCSHIST 1 2025-01-01 "" "User commands"
.ds Nm rcshist
.SH NAME
rcshist \-
display RCS change history
.SH SYNOPSIS
\fB\*(Nm \fI[\fB-mR\fI] [\fB-r\fI branch|\fBMAIN\fI|\fBALL\fI] file ...\fP
.br
\fB\*(Nm -L \fIrevision\fR \fIrcsfile\fR
.SH DESCRIPTION
The \*(Nm utility displays the complete revision history of a set of RCS files
including log messages and patches.
The output is sorted in reverse date order over all revisions of all files.
.PP
The second form displays the patch associated with a particular
revision of an RCS file.
In this case, the revision must be specified numerically and
.I rcsfile
must be an RCS
.B ,v
file.
.PP
The options are as follows:
.IP \fB\-m\fR
Reduce memory usage by retaining only a small fraction of revisions in
memory.
Normally,
\*(Nm
will cache all revisions of all files, since this reduces computation
time significantly.
For very large file sets, this behavior can cause excessive memory usage.
.IP \fB\-R\fR
Recursively search all paths specified for files to analyze.
.IP "\fB\-r\fR \fIbranch|MAIN|ALL\fR"
Restrict output to revisions on the specified symbolic branch tag.
Two special branch names are supported:
.RS
.IP MAIN
Only revisions on the main trunk branch are displayed.
.IP ALL
All revisions are displayed regardless of their branch.
This is the default if \*(Nm cannot infer a branch tag from a CVS/Tag file.
.RE
.PP
Each
.I file
listed on the command line specifies an RCS file to be examined.
The filename is interpreted as follows:
.PP
If the file has a
.B ,v
suffix,
\*(Nm considers it as an RCS format file and opens it directly.
.PP
If the file has no
.B ,v
suffix,
\*(Nm searches for a CVS directory in the same directory as the file.
If this exists,
\*(Nm
attempts to find the RCS file based on the contents of the
CVS/Root and CVS/Repository files.
If a CVS/Tag file exists, and if no branch tag has been specified
on the command-line or inferred from elsewhere, then
\*(Nm
uses the branch tag from CVS/Tag as the default branch to
display.
.PP
Finally, if the file has no
.B ,v
suffix and no CVS directory was found,
\*(Nm
searches for an RCS directory in the same directory as the file.
If this exists,
\*(Nm
will look for an RCS
.B ,v
file corresponding to the specified file.
.PP
If \*(Nm fails to locate an RCS file
corresponding to a specified command-line filename,
\*(Nm
will simply print a warning message and continue.
.SH ENVIRONMENT
.IP RCS_DIR
if defined, specifies the directory in which RCS archive files are found.
Normally files are found in "./RCS".
.SH AUTHORS
Ian Dowse <iedowse@FreeBSD.org>
.br
Thomas E. Dickey <dickey@invisible-island.net>
.SH SEE ALSO
.BR checkout (1),
.BR cvs (1),
.BR rcs (1),
.BR rcsdiff (1),
.BR rcsfile (5),
.BR rlog (1)


CHANGES
2025-01-01  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/copyright, COPYING: bump

	* rcshist.h: RCS_BASE

	* strbuf.c, rcshist.c, rcsfile.c, namedobjlist.c, misc.c:
	aborting on a fatal error is unfriendly - tell the user where it died,
	and allow stdout to be flushed, first

	* rcshist.c:
	ensure a space between filename and date in the "REV:" lines (report by
	Manfred Schwarb)

	* rcshist.1: document $RCS_DIR

	* rcshist.1: convert from mdoc to man, for portability

	* rcsfile.c:
	support $RCS_DIR for compatibility with cm_tools utilities.

	* rcshist.c, misc.c, strbuf.c, namedobjlist.c, namedobjlist.h, rcsfile.c:
	gcc15-warnings

	* configure: regen

	* configure.in: put --with-systype up front

	* aclocal.m4: resync with my-autoconf

	* package/rcshist.spec, package/debian/changelog: bump

2024-07-28  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: 2024-07-27

2024-06-03  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.sub: 2024-05-27

2024-04-19  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/changelog, package/rcshist.spec: bump

	* namedobjlist.c, rcshist.c, strbuf.c, rcsfile.c, misc.c: standard-C

	* package/debian/copyright, COPYING: bump

	* configure: regen

2023-12-27  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.sub: 2023-12-27

2023-12-02  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: 2023-12-02

2023-11-23  Thomas E. Dickey  <dickey@invisible-island.net>

	* aclocal.m4: resync with my-autoconf

2023-01-30  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/copyright, package/debian/changelog,
	  package/rcshist.spec:
	bump

2023-01-29  Thomas E. Dickey  <dickey@invisible-island.net>

	* aclocal.m4: resync with my-autoconf

	* COPYING: bump

	* configure: regen

2022-10-02  Thomas E. Dickey  <dickey@invisible-island.net>

	* config_h.in: s/Header/Id/

	* package/debian/rules, package/debian/compat: fix-compat

	* package/debian/copyright, COPYING: bump

	* configure: regen

	* aclocal.m4, configure.in: resync with my-autoconf

	* package/debian/changelog, package/rcshist.spec: bump

2022-08-01  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess, config.sub: 2022-08-01

2019-01-06  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/copyright, aclocal.m4:
	use the same wording for copyright notice as the original code

	* COPYING: RCS_BASE

	* makefile.in: add dummy rule "check" to appease Arch mkpkg

	* package/debian/changelog: fix typo

	* package/debian/copyright: bump

	* package/rcshist.spec: use hardening flags

	* package/rcshist.spec, package/debian/changelog: bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

2019-01-04  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.sub: 2019-01-05

2019-01-02  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: 2019-01-03

2018-03-25  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/copyright: update/correct

	* package/debian/watch: update ftp-url

	* package/debian/changelog: bump

	* package/debian/compat: update compat to 7

	* package/debian/rules: use Debian build-flags
	enable dh_installdocs, for copyright

	* package/debian/control:
	change to "optional", since "extra" is becoming deprecated

	* namedobjlist.h: fix a pedantic warning with gcc 4.9.4

	* strbuf.c: fix a fall-through warning with gcc 6

	* package/rcshist.spec: update ftp-url
	disable debug-package

	* configure: regen

	* makefile.in: cleanup the new temp-file

	* aclocal.m4, configure.in:
	updated CF_MAKE_DOCS relies upon --with-man2html option (add that)

	* configure: regen

	* aclocal.m4: resync with my-autoconf

2018-02-05  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.sub: 2018-01-15

	* config.guess: 2018-01-26

2015-03-01  Thomas E. Dickey  <dickey@invisible-island.net>

	* rcsfile.c, namedobjlist.c, namedobjlist.h: gcc warnings

	* configure: regen

	* configure.in: autogenerated config.h

	* rcsfile.c, rcsfile.h:
	Add parsing for "commitid", having acquired one in a file using rcs on FreeBSD.
	That came into cvs around 2005 as seen here:
		http://web.mit.edu/andersk/src/cvs-1.12.13.1/contrib/rcs-5.7-commitid.patch

	* package/debian/changelog, package/rcshist.spec:
	*** empty log message ***

	* strbuf.c, misc.c, misc.h, rcsfile.c, namedobjlist.c, rcshist.c:
	gcc warnings

2015-02-28  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/rcshist.spec, package/debian/rules, package/debian/copyright,
	  package/debian/control, package/debian/changelog, makefile.in,
	  configure, aclocal.m4, configure.in:
	RCS_BASE

2014-10-06  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/source/format: RCS_BASE

2014-09-01  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.sub: RCS_BASE

2014-04-19  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: RCS_BASE

2010-04-20  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/compat, package/debian/docs, package/debian/postinst,
	  package/debian/prerm, package/debian/watch:
	RCS_BASE

2005-02-07  iedowse

	* Makefile, rcshist.c, rcshist.1: rcshist-1.04

2004-09-03  iedowse

	* Makefile, rcsfile.c: rcshist-1.03

2004-05-30  iedowse

	* Makefile, strbuf.c: rcshist-1.02

2003-11-29  Thomas E. Dickey  <dickey@invisible-island.net>

	* install-sh: RCS_BASE

2003-10-30  iedowse

	* namedobjlist.c, rcshist.c: rcshist 1.01

	* Makefile, rcsfile.c: rcshist-1.01

2002-02-27  iedowse

	* Makefile, misc.c, bsd_queue.h, rcshist.1, rcshist.c, rcsfile.c:
	rcshist-1.0

2001-05-20  iedowse

	* misc.h, namedobjlist.c, namedobjlist.h, rcsfile.h, strbuf.c, strbuf.h:
	rcshist-1.0

1994-07-11  Thomas E. Dickey  <dickey@clark.net>

	* config_h.in: RCS_BASE



rcshist.c
/*
 * Copyright (c) 2001 Ian Dowse <iedowse@maths.tcd.ie>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer
 *    in this position and unchanged.
 * 2. The name of the author may not be used to endorse or promote products
 *    derived from this software withough specific prior written permission
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * $Id: rcshist.c,v 1.17 2025/01/01 23:44:15 tom Exp $
 */
#include <sys/types.h>
#include <sys/stat.h>
#include <err.h>
#include <fts.h>
#include <unistd.h>

#include "rcshist.h"
#include "namedobjlist.h"
#include "rcsfile.h"
#include "misc.h"

void filelist_expand(char ***filelistp, int *nfilesp);
int filelist_ftscmp(const FTSENT * *fe1, const FTSENT * *fe2);
void prrev(struct revnode *revp);
void prlist(const char *prefix, struct textlist *tlp);
void prlog(struct revnode *revp);
void onerev(char *filename, char *revame);

char *progname;
int mflag;

void
give_up(const char *fn, int ln)
{
	fflush(stdout);
	fprintf(stderr, "%s: fatal error at %s line %d\n", progname, fn, ln);
	exit(EXIT_FAILURE);
}

static void
usage(void) {
	fprintf(stderr,
	    "Usage: %s [-mR] [-r<branch|MAIN|ALL>] <filename> ...\n"
	    "       %s -L<revision> <filename>\n",
	    progname, progname);
	exit(1);
}

int
main(int argc, char **argv) {
	struct rcsfile **rcsp;
	int ch, i, nfiles;
	char *branch = NULL;
	char *revname = NULL;
	char **filelist;
	struct revnode **rlist, **rltmp;
	int Rflag, rnum, rlist_len;

	progname = argv[0];
	Rflag = 0;
	while ((ch = getopt(argc, argv, "L:mr:R")) != -1) {
		switch (ch) {
		case 'L':
			revname = optarg;
			break;
		case 'm':
			mflag = 1;
			break;
		case 'r':
			branch = optarg;
			break;
		case 'R':
			Rflag = 1;
			break;
		case '?':
		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;

	if (argc == 0)
		usage();

	nfiles = argc;
	filelist = argv;

	if (revname != NULL) {
		onerev(filelist[0], revname);
		exit(0);
	}

	if (Rflag)
		filelist_expand(&filelist, &nfiles);

	rlist = NULL;
	rnum = 0;
	rlist_len = 0;

	rcsp = malloc((size_t) nfiles * sizeof(*rcsp));
	for (i = 0; i < nfiles; i++) {
		struct revnode **rpp;

		if ((rcsp[i] = rcsfile_smartopen(filelist[i], &branch)) == NULL)
			continue;
		if (mflag)
			rcsfile_setflags(rcsp[i], RCSFILE_LOWMEM);

		if ((rltmp = revlist(rcsp[i], branch)) == NULL) {
			warnx("%s: %s: no such branch\n", filelist[i],
			    argv[2]);
			continue;
		}

		for (rpp = rltmp; *rpp != NULL; rpp++) {
			if (rnum == rlist_len) {
				rlist_len += rlist_len + 1;
				rlist = realloc(rlist, (size_t) rlist_len *
				    sizeof(*rlist));
			}
			rlist[rnum++] = *rpp;
		}
		free(rltmp);
	}

	qsort(rlist, (size_t) rnum, sizeof(*rlist), revbydate);
	for (i = 0; i < rnum; i++)
		prrev(rlist[i]);
	free(rlist);

	for (i = 0; i < nfiles; i++)
		if (rcsp[i] != NULL)
			rcsfile_free(rcsp[i]);

	return 0;
}

void
prrev(struct revnode *revp) {

	printf("REV:%-20.*s%-20.*s %d/%02d/%02d %02d:%02d:%02d       %.*s\n",
	    revp->revtext.len, revp->revtext.start,
	    revp->rcsp->shortfname.len, revp->rcsp->shortfname.start,
	    revp->date.num[0], revp->date.num[1], revp->date.num[2],
	    revp->date.num[3], revp->date.num[4], revp->date.num[5],
	    revp->author.len, revp->author.start);

	prlist("branchpoints:", revp->branchpoints);
	prlist("branches:    ", revp->branches);
	prlist("tags:        ", revp->tags);

	printf("\n");
	prlog(revp);
	printf("\n");
	rev_calc(revp);

#if 0
	TEXTLIST_FOREACH(revp->outputlines, textp)
		printf("%.*s", textp->len, textp->start);
#endif
	rev_diff(revp, 3, 0);
}

void
onerev(char *filename, char *revname) {
	struct rcsfile *rcsp;
	struct revnode *revp;

	if ((rcsp = rcsfile_open(filename)) == NULL)
		err(1, "%s: rcsfile_open", filename);

	revp = namedobjlist_lookup(rcsp->revs, revname, (int) strlen(revname));
	if (revp == NULL)
		errx(1, "%s: %s: revision not found", filename, revname);

	prlist("branchpoints:", revp->branchpoints);
	prlist("branches:    ", revp->branches);
	prlist("tags:        ", revp->tags);
	rev_diff(revp, 3, 0);
}

void
prlist(const char *prefix, struct textlist *tlp) {
	struct rcstext *textp;
	int len = 0;
	int prefixlen = (int) strlen(prefix) + 4;

	if (tlp->len == 0)
		return;

	TEXTLIST_FOREACH(tlp, textp) {
		if (len == 0 || len + textp->len > 75) {
			if (len == 0)
				printf("%-*s", prefixlen, prefix);
			else
				printf(",\n%-*s", prefixlen, "");
			len = prefixlen;
		}
		if (len > prefixlen) {
			printf(", ");
			len += 2;
		}
		printf("%.*s", textp->len, textp->start);
		len += textp->len;
	}
	printf("\n");
}

void
prlog(struct revnode *revp) {
	struct rcstext *textp;
	struct textlist *tlp = textsplit(&revp->log);

	TEXTLIST_FOREACH(tlp, textp) {
		if (textp->len != 1 || textp->start[0] != '\n')
			printf("   ");
		textprint(textp);
	}

	textlist_destroy(tlp);
}

void
filelist_expand(char ***filelistp, int *nfilesp)
{
	FTS *fts;
	FTSENT *fe;
	char **newlist;
	int newlist_arraysize, nfiles;

	fts = fts_open(*filelistp, FTS_PHYSICAL, filelist_ftscmp);
	if (fts == NULL)
		err(1, "ftsopen");

	newlist = malloc(sizeof(*newlist));
	if (newlist == NULL)
		err(1, "realloc");
	newlist_arraysize = 1;
	nfiles = 0;
	while ((fe = fts_read(fts)) != NULL) {
		switch (fe->fts_info) {
		case FTS_DNR:
		case FTS_ERR:
		case FTS_NS:
			warnx("%s: %s", fe->fts_path, strerror(fe->fts_errno));
			continue;
		case FTS_F:
			break;
		default:
			continue;
		}

		if (nfiles + 1 == newlist_arraysize) {
			newlist_arraysize += newlist_arraysize + 1;
			newlist = realloc(newlist, (size_t) newlist_arraysize *
			    sizeof(*newlist));
			if (newlist == NULL)
				err(1, "realloc");
		}
		newlist[nfiles] = strdup(fe->fts_path);
		if (newlist[nfiles] == NULL)
			err(1, "strdup");
		nfiles++;
	}
	if (fts_close(fts) != 0)
		err(1, "fts_close");

	newlist[nfiles] = NULL;
	*filelistp = newlist;
	*nfilesp = nfiles;
}

int
filelist_ftscmp(const FTSENT * *fe1, const FTSENT * *fe2)
{
	return (strcoll((*fe1)->fts_name, (*fe2)->fts_name));
}
